crates/store/src/db/models/queries/accounts.rs

```diff
@@ -1010,3 +1205,235 @@ pub(crate) struct AccountStorageMapRowInsert {
+///
+/// This should be called by all `_at_block` query functions to ensure that
+/// queries are only performed against blocks that have been produced.
+///
+/// # Arguments
+///
+/// * `conn` - Database connection
+/// * `block_num` - The block number to check
+///
+/// # Returns
+///
+/// * `Ok(())` - If the block exists
+/// * `Err(DatabaseError::BlockNotFound)` - If the block doesn't exist
+/// * `Err(DatabaseError)` - If there's a database error
+fn block_exists(conn: &mut SqliteConnection, block_num: BlockNumber) -> Result<(), DatabaseError> {
+    use schema::block_headers;
+
+    let count: i64 = SelectDsl::select(
+        block_headers::table.filter(block_headers::block_num.eq(block_num.to_raw_sql())),
+        diesel::dsl::count(block_headers::block_num),
+    )
+    .first(conn)?;
+
+    if count > 0 {
+        Ok(())
+    } else {
+        Err(DatabaseError::BlockNotFound(block_num))
+    }
+}
+
+/// Queries the account code for a specific account at a specific block number.
+///
+/// Returns `None` if:
+/// - The account doesn't exist at that block
+/// - The account has no code (private account or account without code commitment)
+///
+/// # Arguments
+///
+/// * `conn` - Database connection
+/// * `account_id` - The account ID to query
+/// * `block_num` - The block number at which to query the account code
+///
+/// # Returns
+///
+/// * `Ok(Some(Vec<u8>))` - The account code bytes if found
+/// * `Ok(None)` - If account doesn't exist or has no code
+/// * `Err(DatabaseError)` - If there's a database error
+pub(crate) fn select_account_code_at_block(
+    conn: &mut SqliteConnection,
+    account_id: AccountId,
+    block_num: BlockNumber,
+) -> Result<Option<Vec<u8>>, DatabaseError> {
+    use schema::{account_codes, accounts};
+
+    // Check if the requested block exists (returns error if not)
+    block_exists(conn, block_num)?;
+
+    let account_id_bytes = account_id.to_bytes();
+    let block_num_sql = i64::from(block_num.as_u32());
+    // Query the accounts table to get the code_commitment at the specified block or earlier
+    // Then join with account_codes to get the actual code
+    let result: Option<Vec<u8>> = SelectDsl::select(
+        accounts::table
+            .inner_join(account_codes::table)
+            .filter(accounts::account_id.eq(&account_id_bytes))
+            .filter(accounts::block_num.le(block_num_sql))
+            .order(accounts::block_num.desc())
+            .limit(1),
+        account_codes::code,
+    )
+    .first(conn)
+    .optional()?;
+
+    Ok(result)
+}
+
+/// Queries the account header for a specific account at a specific block number.
+///
+/// This reconstructs the `AccountHeader` by joining multiple tables:
+/// - `accounts` table for `account_id`, `nonce`, `code_commitment`
+/// - `account_vault_headers` table for `vault_root`
+/// - `account_storage_headers` table for storage slot commitments (to compute `storage_commitment`)
```

Comment:
Related to some of the above comments: ideally, we'd be able to read the full account header just from the `accounts`. This should simplify queries like this - and I'm not seeing downsides to doing this.